### PR TITLE
[oneMKL][Sparse BLAS] Add a new `oneapi::mkl::sparse::set_coo_data()` API for COO matrices

### PR DESCRIPTION
**Pull request by gajanan-choudhary**
## Coordinate (COO) matrix format: `oneapi::mkl::sparse::set_coo_data()` API
* We propose adding a new `oneapi::mkl::sparse::set_coo_data()` API to the oneMKL Specification in the Sparse BLAS domain for creation of matrices in the coordinate (COO) format.
* The API looks very similar to the existing `oneapi::mkl::sparse::set_csr_data()` API in the Spec for the compressed sparse row (CSR) matrix format.
* Proposed `oneapi::mkl::sparse::set_coo_data()` API
    ```cpp
    namespace oneapi::mkl::sparse {
        /* sycl::buffer API */
        void set_coo_data(sycl::queue                           &queue,
                          oneapi::mkl::sparse::matrix_handle_t  handle,
                          const intType                         nrows,
                          const intType                         ncols,
                          const intType                         nnz,
                          oneapi::mkl::index_base               index,
                          sycl::buffer<intType, 1>              &row_ind,
                          sycl::buffer<intType, 1>              &col_ind,
                          sycl::buffer<fp, 1>                   &val);
        /* USM API */
        sycl::event set_coo_data(sycl::queue                           &queue,
                                 oneapi::mkl::sparse::matrix_handle_t  handle,
                                 const intType                         nrows,
                                 const intType                         ncols,
                                 const intType                         nnz,
                                 oneapi::mkl::index_base               index,
                                 intType                               *row_ind,
                                 intType                               *col_ind,
                                 fp                                    *val,
                                 const std::vector<sycl::event>        &dependencies = {});
    }
    ```
* Existing`oneapi::mkl::sparse::set_csr_data()` API
    ```cpp
    namespace oneapi::mkl::sparse {
        /* sycl::buffer API */
        void set_csr_data(sycl::queue                           &queue,
                          oneapi::mkl::sparse::matrix_handle_t  handle,
                          const intType                         nrows,
                          const intType                         ncols,
                          const intType                         nnz,
                          oneapi::mkl::index_base               index,
                          sycl::buffer<intType, 1>              &row_ptr,
                          sycl::buffer<intType, 1>              &col_ind,
                          sycl::buffer<fp, 1>                   &val);
        /* USM API */
        sycl::event set_csr_data(sycl::queue                           &queue,
                                 oneapi::mkl::sparse::matrix_handle_t  handle,
                                 const intType                         nrows,
                                 const intType                         ncols,
                                 const intType                         nnz,
                                 oneapi::mkl::index_base               index,
                                 intType                               *row_ptr,
                                 intType                               *col_ind,
                                 fp                                    *val,
                                 const std::vector<sycl::event>        &dependencies = {});
    }
    ```

## Changes
* A new file, [setcoostructure.rst](https://github.com/gajanan-choudhary/oneAPI-spec/blob/oneMKL/SpBLAS/gajanan-choudhary/add-new-sparse-set_coo_data-API/source/elements/oneMKL/source/domains/spblas/setcoostructure.rst) is being added containing details of the new `oneapi::mkl::sparse::set_coo_data()` API.
* In almost all the existing sparse BLAS APIs, the description of the sparse matrix handle mentioned the CSR format; that is now being changed to reflect both CSR and COO formats.
* [format-descriptions.rst](https://github.com/gajanan-choudhary/oneAPI-spec/blob/oneMKL/SpBLAS/gajanan-choudhary/add-new-sparse-set_coo_data-API/source/elements/oneMKL/source/domains/spblas/format-descriptions.rst) is modified to also contain a description and example of the COO format. The COO format description is kept at the top since it is the simplest one to understand.
* [Update:] After a PR review, I've added in text for clarifying that `set_csr_data` and `set_coo_data` APIs need to allow multiple calls and `0`/`nullptr` args in some cases and that it is implementation-defined for now.
* Remaining changes, if any, are very minor.

## Tests
* Built the HTML pages locally and checked each Sparse BLAS page for rendering quality, which looks normal.
* Tested COO format related links in all webpages.
* As a sanity check, there are currently exactly 46 lines each with (case-insensitive) occurrences of the words, "CSR" and "COO", in the sparse BLAS domain files.

### CODE REVIEW

File: source/elements/oneMKL/source/domains/spblas/format-descriptions.rst [hunk 1]
```diff
@@ -7,16 +7,87 @@
...
++------------+------------------------------------------------------------+
+| nnz        | 5                                                          |
++------------+------------------------------------------------------------+
+| index      | 1                                                          |
++------------+------------+-----------+-----------+-----------+-----------+
+| row_ind    | 1          | 1         | 2         | 2         | 3         |
++------------+------------+-----------+-----------+-----------+-----------+
+| col_ind    | 1          | 3         | 2         | 3         | 1         |
++------------+------------+-----------+-----------+-----------+-----------+
+| val        | 1          | 2         | -1        | 4         | 3         |
```
Comment (noffermans) [inline source/elements/oneMKL/source/domains/spblas/format-descriptions.rst:78]: Just a minor comment, in the CSR description, we present the arrays in the following order: `val` - `col_ind` - `row_ptr`. Do we want to follow the same order here? Or change it for CSR?
Or maybe each format should be presented in its own way that feels more "natural".

Comment (noffermans) [reply source/elements/oneMKL/source/domains/spblas/format-descriptions.rst:78]: Actually, I'd rather change the CSR description to match the order of the arguments in the spec. This looks good for the COO format.

**Review by noffermans — APPROVED** [review]
Looks good to me!

**Review by spencerpatty — APPROVED** [review]
changes look fine :) approved

File: source/elements/oneMKL/source/domains/spblas/format-descriptions.rst (hunk 1 above)
Comment (gajanan-choudhary) [reply source/elements/oneMKL/source/domains/spblas/format-descriptions.rst:78]: I did think of this actually. Eventually I kept CSR unchanged for minimizing the changes. I can make this change.

Comment (noffermans) [reply source/elements/oneMKL/source/domains/spblas/format-descriptions.rst:78]: If you think it's worth it. I'm fine if we keep it as is too

Comment (gajanan-choudhary) [reply source/elements/oneMKL/source/domains/spblas/format-descriptions.rst:78]: I've made this change now.

Comment (noffermans) [reply source/elements/oneMKL/source/domains/spblas/format-descriptions.rst:78]: Thanks!

**Pull request merged**